src/librustc/mir/tcx.rs

```diff
@@ -151,12 +153,124 @@ impl<'tcx> Place<'tcx> {
+        match self {
+            PlaceBase::Local(index) => local_decls.local_decls()[*index].ty,
+            PlaceBase::Promoted(data) => data.1,
+            PlaceBase::Static(data) => data.ty,
+        }
+    }
+}
+
+impl<'tcx> NeoPlace<'tcx> {
+    pub fn ty<'a, 'gcx>(
+        &self,
+        local_decls: &impl HasLocalDecls<'tcx>,
+        tcx: TyCtxt<'a, 'gcx, 'tcx>,
+    ) -> PlaceTy<'tcx> {
+        // the PlaceTy is the *final* type with all projection applied
+        // if there is no projection, that just refers to `base`:
+        //
+        // Place: base.[a, b, c]
+        //                    ^-- projection
+        //                    ^-- PlaceTy
+        //
+        // Place: base.[]
+        //             ^^-- no projection
+        //        ^^^^-- PlaceTy
+
+        let mut place_ty = PlaceTy::from(self.base.ty(local_decls));
+
+        // apply .projection_ty() to all elems but only returns the final one.
+        for elem in self.elems.iter() {
+            place_ty = place_ty.projection_ty(tcx, elem);
+        }
+
+        place_ty
+    }
+
+    /// If this is a field projection, and the field is being projected from a closure type,
+    /// then returns the index of the field being projected. Note that this closure will always
+    /// be `self` in the current MIR, because that is the only time we directly access the fields
+    /// of a closure type.
+    pub fn is_upvar_field_projection<'cx, 'gcx>(
+        &self,
+        mir: &'cx Mir<'tcx>,
+        tcx: &TyCtxt<'cx, 'gcx, 'tcx>,
+    ) -> Option<Field> {
+        let (place, by_ref) = if let Some(ProjectionElem::Deref) = self.elems.last() {
+            (self.projection_base(*tcx), true)
+        } else {
+            (self.clone(), false)
+        };
+
+        if let Some(ProjectionElem::Field(field, _ty)) = place.elems.last() {
+            let base_ty = place.projection_base(*tcx).ty(mir, *tcx).to_ty(*tcx);
+            if (base_ty.is_closure() || base_ty.is_generator()) &&
+                (!by_ref || mir.upvar_decls[field.index()].by_ref)
+            {
+                Some(*field)
+            } else {
+                None
+            }
+        } else {
+            None
+        }
+    }
+
+    // for Place:
+    //    (Base.[a, b, c])
+    //     ^^^^^^^^^^  ^-- projection
+    //     |-- base_place
+    //
+    //     Base.[]
+    //     ^^^^ ^^-- no projection(empty)
+    //     |-- base_place
+    pub fn split_projection<'cx, 'gcx>(
+        &self,
+        tcx: TyCtxt<'cx, 'gcx, 'tcx>,
+    ) -> (NeoPlace<'tcx>, Option<&'tcx PlaceElem<'tcx>>) {
+        // split place_elems
+        // Base.[a, b, c]
+        //       ^^^^  ^-- projection(projection lives in the last elem)
+        //       |-- place_elems
+        match self.elems.split_last() {
+            Some((projection, place_elems)) => (
+                NeoPlace {
+                    base: self.clone().base,
+                    elems: tcx.intern_place_elems(place_elems),
```

Comment:
There is no need to intern again here; just use `place_elems` directly. It is already pointing into interned memory, right?